cip/1.accepted/CIP2016-12-14-Constraint-syntax.adoc

```diff
@@ -0,0 +1,350 @@
+CREATE (:Color {name: 'black', rgb: 0x000000})
+CREATE (:Color {name: 'very, very dark grey', rgb: 0x000000}) // rounding error!
+----
+
+Owing to the duplication of the `rgb` property, the following attempt at creating a constraint will fail:
+
+[source, cypher]
+----
+CREATE CONSTRAINT only_one_color_per_rgb
+FOR (c:Color)
+REQUIRE UNIQUE c.rgb
+----
+
+Suppose that we would rather like to have one color node per `name` _and_ `rgb` value (to work around the rounding errors).
+We could then use the following constraint, without modifying our data:
+
+[source, cypher]
+----
+CREATE CONSTRAINT unique_color_nodes
+FOR (c:Color)
+REQUIRE UNIQUE c.rgb
+REQUIRE UNIQUE c.name
+----
+
+Now, consider the following query which retrieves the RGB value of a color with a given `name`:
+
+[source, cypher]
+----
+MATCH (c:Color {name: $name})
+WHERE exists(c.rgb)
+RETURN c.rgb
+----
+
+The `WHERE` clause is here used to prevent an application from retrieving `null` values for user-defined colors where the RGB values have not been specified correctly.
+It may, however, be eliminated by the introduction of a constraint asserting the existence of that property:
+
+[source, cypher]
+----
+CREATE CONSTRAINT colors_must_have_rgb
+FOR (c:Color)
+REQUIRE exists(c.rgb)
+----
+
+Any updating statement that would create a `:Color` node without specifying an `rgb` property for it would now fail.
+
+If we also want to mandate the existence of the `name` property, we could use a `PRIMARY KEY` operator to capture all these requirements in a single constraint:
+
+[source, cypher]
+----
+CREATE CONSTRAINT color_schema
+FOR (c:Color)
+REQUIRE PRIMARY KEY c.rgb, c.name
+----
+
+This constraint will make sure that all `:Color` nodes has a value for their `rgb` and `name` properties, and that the combination is unique across all the nodes.
+
+More complex constraint definitions are considered below:
+
+.Multiple property existence using conjunction
+[source, cypher]
+----
+CREATE CONSTRAINT person_properties
+FOR (p:Person)
+REQUIRE exists(p.name) AND exists(p.email)
+----
+
+.Using larger pattern
+[source, cypher]
+----
+CREATE CONSTRAINT not_rating_own_posts
+FOR (u1:User)-[:RATED]->(:Post)<-[:POSTED_BY]-(u2:User)
+REQUIRE u.name <> u2.name
+----
+
+.Property value limitations
+[source, cypher]
+----
+CREATE CONSTRAINT road_width
+FOR ()-[r:ROAD]-()
+REQUIRE 5 < r.width < 50
+----
+
+.Cardinality
+[source, cypher]
+----
+CREATE CONSTRAINT spread_the_love
+FOR (p:Person)
+REQUIRE size((p)-[:LOVES]->()) > 3
+----
+
+.Endpoint requirements
+[source, cypher]
+----
+CREATE CONSTRAINT can_only_own_things
+FOR ()-[:OWNS]->(t)
+REQUIRE (t:Vehicle) OR (t:Building) OR (t:Object)
+----
+
+.Label coexistence
+[source, cypher]
+----
+CREATE CONSTRAINT programmers_are_people_too
+FOR (p:Programmer)
+REQUIRE p:Person
+----
+
+.Constraint example from CIR-2017-173
+[source, cypher]
+----
+CREATE CONSTRAINT enforce_dag_unique_for_R_links
+FOR p = (a)-[:R]-(b)
+REQUIRE length(p) < 2
+----
```

Comment:
You're right; `length()` actually doesn't make much sense here (it will always be 1 with this pattern). Although `size()` doesn't either, as it's not defined over paths at all.

Perhaps 

```
REQUIRE size(collect(DISTINCT p)) < 2
```

would work, but it's not very pretty. We probably need some other construct here. I'll remove this example for now.